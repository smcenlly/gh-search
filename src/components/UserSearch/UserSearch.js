import React from "react";
import i18n from "i18next";
import styled from "styled-components";
import PropTypes from "prop-types";
import { Button } from "../../toolkit";
import profilePicUrl from "./assets/profile.svg";
import { colors } from "../../styles";

const UserForm = styled.form`
    display: flex;
    flex-shrink: 0;
    padding: 20px;
    border-bottom: 1px solid ${colors.separator};
`;

const Img = styled.img.attrs({
    src: profilePicUrl,
    width: "15px",
    height: "15px"
})`
    vertical-align: middle;
`;

const Label = styled.label.attrs({
    htmlFor: "UserSearch"
})`
    vertical-align: middle;
    margin-left: 5px;
`;

const UserFormItem = styled.div`
    & + & {
        margin-left: 12px;
    }
`;

const SearchInput = styled.input`
    padding: 2px 5px;
    background: none;
    outline: none;
    border-radius: 3px;
    border: 1px solid ${colors.separator};
`;

class UserSearch extends React.PureComponent {
    static propTypes = {
        onChange: PropTypes.func,
        onSubmit: PropTypes.func,
        value: PropTypes.string,
        isSubmitting: PropTypes.bool
    };

    static defaultProps = {
        value: "",
        isSubmitting: false
    };

    static SearchInput = SearchInput;
    static SearchButton = Button;

    handleSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit && this.props.onSubmit(e);
    };

    isValid = () => {
        return this.props.value && !this.props.isSubmitting;
    };

    render() {
        return (
            <UserForm
                className={this.props.className}
                onSubmit={this.handleSubmit}
            >
                <UserFormItem>
                    <Img />
                    <Label>{i18n.t("userSearch.label")}</Label>
                </UserFormItem>
                <UserFormItem>
                    <SearchInput
                        id="UserSearch"
                        onChange={this.props.onChange}
                        value={this.props.value}
                    />
                </UserFormItem>
                <UserFormItem>
                    <Button disabled={!this.isValid()} type="submit">
                        {i18n.t("userSearch.submit")}
                    </Button>
                </UserFormItem>
            </UserForm>
        );
    }
}

export default UserSearch;
