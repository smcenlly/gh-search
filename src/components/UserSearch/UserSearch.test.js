import React from "react";
import { shallow } from "enzyme";
import UserSearch from "./UserSearch";

describe("UserSearch", () => {
    let wrapper,
        getSearchButton = () => wrapper.find(UserSearch.SearchButton),
        getSearchInput = () => wrapper.find(UserSearch.SearchInput);

    beforeEach(() => {
        wrapper = shallow(<UserSearch />);
    });

    it("renders correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("initializes empty", () => {
        expect(wrapper.instance().props).toEqual({
            value: "",
            isSubmitting: false
        });
    });

    it("renders SearchInput with correct props", () => {
        const props = { onChange: () => {}, value: "test" };

        wrapper.setProps(props);

        expect(wrapper).toContainExactlyOneMatchingElement(
            UserSearch.SearchInput
        );
        expect(wrapper.find(UserSearch.SearchInput).first()).toHaveProp(props);
    });

    it("renders SearchButton", () => {
        expect(wrapper).toContainExactlyOneMatchingElement(
            UserSearch.SearchButton
        );
    });

    describe("when not having value", () => {
        it("disables submit", () => {
            wrapper.setProps({ value: "", isSubmitting: false });
            expect(getSearchButton()).toHaveProp({ disabled: true });
        });
    });

    describe("when is submitting", () => {
        it("disables submit", () => {
            wrapper.setProps({ value: "Something", isSubmitting: true });
            expect(getSearchButton()).toHaveProp({ disabled: true });
        });
    });

    describe("when typing in search input", () => {
        it("calls onChange", () => {
            const onChange = jest.fn();
            wrapper.setProps({ onChange, value: "Tex" });

            const changeEvent = { target: { value: "Text" } };

            getSearchInput().simulate("change", changeEvent);

            expect(onChange).toHaveBeenCalled();
            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange).toHaveBeenCalledWith(changeEvent);
        });
    });

    describe("when valid and submitting", () => {
        it("calls onSubmit", () => {
            const onSubmit = jest.fn();
            wrapper.setProps({ onSubmit, value: "Text", isSubmitting: false });

            const submitEvent = { preventDefault: jest.fn() };

            wrapper.simulate("submit", submitEvent);

            expect(submitEvent.preventDefault).toHaveBeenCalled();
            expect(submitEvent.preventDefault).toHaveBeenCalledTimes(1);
            expect(onSubmit).toHaveBeenCalled();
            expect(onSubmit).toHaveBeenCalledTimes(1);
            expect(onSubmit).toHaveBeenCalledWith(submitEvent);
        });
    });
});
