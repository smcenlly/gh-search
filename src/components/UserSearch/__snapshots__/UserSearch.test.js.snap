// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSearch renders correctly 1`] = `
<styled.form
  onSubmit={[Function]}
>
  <styled.div>
    <styled.img />
    <styled.label />
  </styled.div>
  <styled.div>
    <styled.input
      id="UserSearch"
      value=""
    />
  </styled.div>
  <styled.div>
    <styled.button
      disabled={true}
      type="submit"
    />
  </styled.div>
</styled.form>
`;
